text/0000-debug-map-key-value.md

```diff
@@ -0,0 +1,284 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut map = f.debug_map();
+
+        // Equivalent to map.entry
+        map.key(&"key").value(&"value");
+
+        map.finish()
+    }
+}
+```
+
+Every call to `key` must be directly followed by a corresponding call to `value` to complete the entry:
+
+```rust
+impl fmt::Debug for Map {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut map = f.debug_map();
+
+        map.key(&1);
+
+        // err: attempt to start a new entry without finishing the current one
+        map.key(&2);
+
+        map.finish()
+    }
+}
+```
+
+`key` must be called before `value`:
+
+```rust
+impl fmt::Debug for Map {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut map = f.debug_map();
+
+        // err: attempt to write a value without first writing its key
+        map.value(&"value");
+        map.key(&"key");
+
+        map.finish()
+    }
+}
+```
+
+Each entry must be finished before the map can be finished:
+
+```rust
+impl fmt::Debug for Map {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut map = f.debug_map();
+
+        map.key(&1);
+
+        // err: attempt to finish a map that has an incomplete key
+        map.finish()
+    }
+}
+```
+
+Any incorrect calls to `key` and `value` will poison the `DebugMap` and cause it to return an error when calling `finish`.
+
+## When to use `key` and `value`
+
+Why would you want to use `key` and `value` directly if they're less convenient than `entry`? The reason is when the driver of the `DebugMap` is a framework like `serde` rather than a data structure directly:
+
+```rust
+struct DebugMap<'a, 'b: 'a>(fmt::DebugMap<'a, 'b>);
+
+impl<'a, 'b: 'a> SerializeMap for DebugMap<'a, 'b> {
+    type Ok = ();
+    type Error = Error;
+
+    fn serialize_key<T: ?Sized>(&mut self, key: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.0.key(&key.to_debug());
+        Ok(())
+    }
+
+    fn serialize_value<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
+    where
+        T: Serialize,
+    {
+        self.0.value(&value.to_debug());
+        Ok(())
+    }
+
+    fn serialize_entry<K: ?Sized, V: ?Sized>(
+        &mut self,
+        key: &K,
+        value: &V,
+    ) -> Result<(), Self::Error>
+    where
+        K: Serialize,
+        V: Serialize,
+    {
+        self.0.entry(&key.to_debug(), &value.to_debug());
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        self.0.finish().map_err(Into::into)
+    }
+}
+```
+
+Consumers should prefer calling `entry` over `key` and `value`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `key` and `value` methods can be implemented on `DebugMap` by tracking the state of the current entry in a `bool`, and splitting the existing `entry` method into two:
+
+```rust
+pub struct DebugMap<'a, 'b: 'a> {
```

Comment:
Note that the code presented here is actually incorrect. I’ll probably just remove it in favour of some prose + a correct implementation in the PR.